abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,2166 @@
+- *Stream Context* - a context in which operations from a specific HTTP
+  request/response pair or TCP stream are executed.
+
+
+# Callbacks and Functions
+
+### Callbacks exposed by the Wasm module
+
+All callbacks (entry points named `proxy_on_<event>`) are optional,
+and are going to be called only if exposed by the Wasm module.
+
+All callbacks, other than the [integration] and [memory management]
+functions, include a unique context identifier as the first parameter,
+which can be used to distinguish between different contexts.
+
+
+### Functions exposed by the host
+
+All functions exposed by the host are required.
+
+All Proxy-Wasm functions exposed by the host return [`proxy_status_t`],
+which indicates status of the call (success, invalid memory access,
+etc.). Return values are written into memory pointed by `return_<value>`
+parameters.
+
+
+## Integration
+
+### Callbacks exposed by the Wasm module
+
+#### `proxy_abi_version_0_2_1`
+
+* params:
+  - none
+* returns:
+  - none
+
+Function marker used during linking to advertise Wasm module's support
+for Proxy-Wasm ABI v0.2.1.
+
+This function is never called.
+
+
+#### `_initialize`
+
+* params:
+  - none
+* returns:
+  - none
+
+Called when the Wasm module is first loaded.
+
+
+#### `main`
+
+* params:
+  - `i32 (uint32_t) unused`
+  - `i32 (uint32_t) unused`
+* returns:
+  - `i32 (uint32_t) unused`
+
+> **Warning**
+> This is called only if [`_initialize`] is also exported.
+
+Called when the Wasm module is first loaded, after [`_initialize`].
+
+
+#### `_start`
+
+* params:
+  - none
+* returns:
+  - none
+
+> **Warning**
+> This is called only if [`_initialize`] is not exported.
+
+Called when the Wasm module is first loaded.
+
+
+## Memory management
+
+### Callbacks exposed by the Wasm module
+
+#### `proxy_on_memory_allocate`
+
+* params:
+  - `i32 (size_t) memory_size`
+* returns:
+  - `i32 (uint8_t *) memory_data`
+
+Called to allocate continuous memory buffer of `memory_size` using
+the in-VM memory allocator.
+
+Plugin must return `memory_data` pointing to the start of the allocated
+memory.
+
+Returning `0` indicates failure.
+
+
+#### `malloc`
+
+* params:
+  - `i32 (size_t) memory_size`
+* returns:
+  - `i32 (uint8_t *) memory_data`
+
+> **Warning**
+> This callback has been deprecated in favor of [`proxy_on_memory_allocate`],
+> and it's called only in its absence.
+
+Called to allocate continuous memory buffer of `memory_size` using
+the in-VM memory allocator.
+
+Plugin must return `memory_data` pointing to the start of the allocated
+memory.
+
+Returning `0` indicates failure.
+
+
+## Context lifecycle
+
+### Callbacks exposed by the Wasm module
+
+#### `proxy_on_context_create`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) parent_context_id`
+* returns:
+  - none
+
+Called when the host creates a new context (`context_id`).
+
+When `parent_context_id` is `0` then a new plugin context is created,
+otherwise a new per-stream context is created.
```

Comment:
... in which case the parent_context_id refers to the plugin context.

Reply:
Done.